File: 3714-longest-balanced-substring-ii/3714-longest-balanced-substring-ii.py
def case1(s: str, x: str) -> int:
    res = 0
    i = 0
    while i < len(s):
        cnt = 0
        while i < len(s) and s[i] == x:
            cnt += 1
            res = max(res, cnt)
            i += 1
        i += 1
    return res
def case2(s: str, x: str, y: str) -> int:
    res = 0
    i = 0
    while i < len(s):
        diff = {}
        diff[0] = i - 1
        xCnt = 0
        yCnt = 0
        while i < len(s) and (s[i] == x or s[i] == y):
            if s[i] == x:
                xCnt += 1
            else:
                yCnt += 1
            d = xCnt - yCnt
            if d in diff:
                res = max(res, i - diff[d])
            else:
                diff[d] = i
            i += 1
        i += 1
    return res
def case3(s: str) -> int:
    res = 0
    diff = {}
    a = 0
    b = 0
    c = 0
    d1 = 0
    d2 = 0
    diff[(d1, d2)] = -1
    for i in range(len(s)):
        if s[i] == 'a':
            a += 1
        elif s[i] == 'b':
            b += 1
        else:
            c += 1
        d1 = a - b
        d2 = a - c
        if (d1, d2) in diff:
            res = max(res, i - diff[(d1, d2)])
        else:
            diff[(d1, d2)] = i
    return res 

class Solution:
    def longestBalanced(self, s: str) -> int:
        res = 0
        res = max(res, case1(s, 'a'))
        res = max(res, case1(s, 'b'))
        res = max(res, case1(s, 'c'))
        res = max(res, case2(s, 'a', 'b'))
        res = max(res, case2(s, 'a', 'c'))
        res = max(res, case2(s, 'b', 'c'))
        res = max(res, case3(s))
        return res

File: 3714-longest-balanced-substring-ii/3714-longest-balanced-substring-ii.cpp
class Solution {
public:
    int longestBalanced(string s) {
        int res = 0;
        res = max(res, case1(s, 'a'));
        res = max(res, case1(s, 'b'));
        res = max(res, case1(s, 'c'));
        res = max(res, case2(s, 'a', 'b'));
        res = max(res, case2(s, 'a', 'c'));
        res = max(res, case2(s, 'b', 'c'));
        res = max(res, case3(s));
        return res;
    }
private:
    int case1(string s, char x)
    {
        int res = 0;
        for(int i = 0; i < s.length(); i++)
        {
            int cnt = 0;
            while(i < s.length() && s[i] == x)
            {
                cnt++;
                res = max(res, cnt);
                i++;
            }
        }
        return res;
    }
    int case2(string s, char x, char y)
    {
        int res = 0;
        for(int i = 0; i < s.length(); i++)
        {
            unordered_map<int, int> diff;
            diff[0] = i - 1;
            int xCnt = 0;
            int yCnt = 0;
            while(i < s.length() && (s[i] == x || s[i] == y))
            {
                if(s[i] == x)
                {
                    xCnt++;
                }
                else
                {
                    yCnt++;
                }
                int curDiff = xCnt - yCnt;
                if(diff.count(curDiff))
                {
                    res = max(res, i - diff[curDiff]);
                }
                else
                {
                    diff[curDiff] = i;
                }    
                i++;
            }
        }
        return res;
    }
    int case3(string s)
    {
        int res = 0;
        unordered_map<string, int> diff;
        int a = 0;
        int b = 0;
        int c = 0;
        int d1 = 0;
        int d2 = 0;
        diff["0,0"] = -1;
        for(int i = 0; i < s.length(); i++)
        {
            if(s[i] == 'a')
            {
                a++;
            }
            else if(s[i] == 'b')
            {
                b++;
            }
            else
            {
                c++;
            }
            d1 = a - b;
            d2 = a - c;
            string hash = to_string(d1) + "," + to_string(d2);
            if(diff.count(hash))
            {
                res = max(res, i - diff[hash]);
            }
            else
            {
                diff[hash] = i;
            }
        }
        return res;
    }
};